### ISSUE TITLE
NPE in RScope.createLocalScopes(RContext, Node, Frame) if Frame == null

### ISSUE THREAD

**Issue by numberpi** [oracle/fastr#31]
The API docu of `TruffleLanguage.findLocalScopes(C, Node, Frame)` allows a null-frame as last argument:

> `frame` The current frame the node is in, or null for lexical access when the program is not running, or is not suspended at the node's location.

But a NullPointerException is thrown when calling with a null-frame.
```
Caused by: java.lang.NullPointerException
	at com.oracle.truffle.r.runtime.RArguments.isRFrame(RArguments.java:356)
	at com.oracle.truffle.r.runtime.env.RScope.getEnv(RScope.java:100)
	at com.oracle.truffle.r.runtime.env.RScope.createLocalScopes(RScope.java:118)
	at com.oracle.truffle.r.engine.TruffleRLanguageImpl.findLocalScopes(TruffleRLanguageImpl.java:231)
	at com.oracle.truffle.r.engine.TruffleRLanguageImpl.findLocalScopes(TruffleRLanguageImpl.java:1)
	at com.oracle.truffle.api.TruffleLanguage$Env.findLocalScopes(TruffleLanguage.java:2028)
	at com.oracle.truffle.api.TruffleLanguage$LanguageImpl.findLocalScopes(TruffleLanguage.java:2494)
	at com.oracle.truffle.api.instrumentation.TruffleInstrument$Env.findLocalScopes(TruffleInstrument.java:646)
```

**Comment by steve-s** [comment]
Hello Daniel, 

thank you for reporting this. The unhandled NPE is indeed a bug on our side and we will fix it. The question now is: do you just want FastR to not crash, or do you expect some useful return value? 

The issue is that all the variables in R are created dynamically, so we cannot easily give you anything else than empty set. We could extract the names of the formal arguments and variables that are referenced directly, but I think this would be misleading because some local variables (that are created dynamically) will be omitted and some formal arguments may not be passed at the call-site.

Out of curiosity: could you share with us what kind of tool are you developing?

**Comment by steve-s** [comment]
Hi Daniel, 

the simple fix that prevents `NPE` is merged in 1c0fdf3463b10aeea941e2b1864566360f2fa68b. For now, we will only return the empty set. In the future, we may improve that.

If you wanted a better version now we are also happy to accept contributions. The implementation could traverse the AST from the `RootNode` to find the `FunctionDefinitionNode` and get the formal arguments from there [1]. Further traversing the AST it may look for `RSyntaxCall` nodes for which the function name is "<-" (`getSyntaxLHS()`[2]) and the first argument [3] is the variable assigned into (use `instanceof RSyntaxLookup => RSyntaxLookup#getIdentifier()`). This of course ignores control flow as a local variable may or may not be defined on certain control flow paths. In any case, this would be far from precise, but maybe good enough for code completition.

[1] https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.nodes/src/com/oracle/truffle/r/nodes/function/FunctionDefinitionNode.java#L89
[2] https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/nodes/RSyntaxCall.java#L34
[3] https://github.com/oracle/fastr/blob/master/com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/nodes/RSyntaxCall.java#L38
